task/sensor/src/main.rs

```diff
@@ -0,0 +1,119 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Sensor management
+
+#![no_std]
+#![no_main]
+
+use idol_runtime::{NotificationHandler, RequestError};
+use task_sensor_api::{NoData, Reading, SensorError, SensorId};
+use userlib::*;
+
+// This is only included to determine the number of sensors
+include!(concat!(env!("OUT_DIR"), "/i2c_config.rs"));
+
+use i2c_config::sensors;
+use sensors::NUM_SENSORS;
+
+struct ServerImpl {
+    data: [Reading; NUM_SENSORS],
+    deadline: u64,
+}
+
+const TIMER_MASK: u32 = 1 << 0;
+const TIMER_INTERVAL: u64 = 1000;
+
+impl idl::InOrderSensorImpl for ServerImpl {
+    fn get(
+        &mut self,
+        _: &RecvMessage,
+        id: SensorId,
+    ) -> Result<f32, RequestError<SensorError>> {
+        let index = id.0;
+
+        if index < NUM_SENSORS {
+            match self.data[index] {
+                Reading::None => Err(SensorError::NoReading.into()),
+                Reading::NoData(nodata) => {
+                    let err: SensorError = nodata.into();
+                    Err(err.into())
+                }
+                Reading::Value(reading) => Ok(reading),
+            }
+        } else {
+            Err(SensorError::InvalidSensor.into())
+        }
+    }
+
+    fn post(
+        &mut self,
+        _: &RecvMessage,
+        id: SensorId,
+        value: f32,
+    ) -> Result<(), RequestError<SensorError>> {
+        let index = id.0;
+
+        if index < NUM_SENSORS {
+            self.data[index] = Reading::Value(value);
+            Ok(())
+        } else {
+            Err(SensorError::InvalidSensor.into())
+        }
+    }
+
+    fn nodata(
+        &mut self,
+        _: &RecvMessage,
+        id: SensorId,
+        nodata: NoData,
+    ) -> Result<(), RequestError<SensorError>> {
+        let index = id.0;
+
+        if index < NUM_SENSORS {
+            self.data[index] = Reading::NoData(nodata);
```

Comment:
We should talk about this at a whiteboard (possibly after this gets merged), but -- it's not clear to me that you want to throw away any existing valid data if the sensor flakes.

Reply:
I think this first sensor task is very much a sketch -- I am assuming that we will want to improve it in many ways, including how data is retained/discarded.